tests/test_forward_ref.py

```diff
@@ -876,9 +876,19 @@ def pytest_raises_undefined_types_warning(defining_class_name, missing_type_name
     )
 
 
-def test_undefined_type_raises_warning(create_module):
-    """Test undefined_types_warning is raised when a model declares a variable of type that is missing."""
-
+#   NOTE: the `undefined_types_warning` tests below are "statically parameterized" (i.e. have Duplicate Code).
+#   The initial attempt to refactor them into a single parameterized test was not strateforward, due to the use of the
+#   `create_module` fixture and the need for `from __future__ import annotations` be the first line in a module.
+#
+#   Test Parameters:
+#     1. config setting: (1a) default behavior vs (1b) overriding with Config setting:
+#     2. type checking approach: (2a) `from __future__ import annotations` vs (2b) `ForwardRef`
+#
+#   The parameter tags "1a", "1b", "2a", and "2b" are used in the test names below, to indicate which combination
+#   of conditions the test is covering.
```

Comment:
Probably the best inline documentation I've seen in a very long time. Well done.

Reply:
thanks! I'll mark this as resolved